### ISSUE TITLE
Monitoring download progress.

### ISSUE THREAD

**Issue by tomchristie** [encode/httpx#1208]
For clients that would like to display something like a progress bar for downloads, there's an issue in that while you're streaming data from `iter_content()`, you *can't* properly know how far you have progressed through the download because we'll often be performing gzip decompression. The `Content-Length` header gives the length of the raw data. (And we really *don't* want to be streaming from `iter_raw()` because we really *do* want the uncompressed bytes, not the bytes on the wire.)

See also for example, this question related to `requests`... https://stackoverflow.com/questions/21778885/python3-progress-bar-and-download-with-gzip

It might make sense to provide an API for monitoring the number of bytes that have been consumed by the underlying `iter_raw` calls, for example something like... `response.num_bytes_downloaded`.

```python
with open("download.txt", "wb") as download_file:
    with httpx.stream('http://www.example.org') as response:
        content_length = response.headers.get("Content-Length")
        for chunk in response.iter_content():
            download_file.write(chunk)
            if content_length is not None:
                percent_complete = 100.0 * (response.num_bytes_downloaded / content_length)
                print(f"{percent_complete:.2f}%")
```

Perhaps we ought to provide a `response.percent_complete` that wraps this up nicely, returning either a 0.0 <= float <= 100.0 or `None` if the `Content-Length` is unknown, so...

```python
with open("download.txt", "wb") as download_file:
    with httpx.stream('http://www.example.org') as response:
        for chunk in response.iter_content():
            download_file.write(chunk)
            if response.percent_complete is not None:
                print(f"{response.percent_complete:.2f}%")
```

**Comment by florimondmanca** [comment]
Regardless of the API we decide on, one way to challenge it and make sure it's right enough for most use cases is to see if it would integrate well with [tqdm](https://github.com/tqdm/tqdm)...

And, well, the way I see it there isn't really a need for us to add anything, since `tqdm` supports a manual [`.update()`](https://tqdm.github.io/docs/tqdm/#update)-based usage that was specifically designed for this kind of "read from a known-length stream" use cases.

For example:

```python
import httpx
from tqdm import tqdm

with open("download.txt", "wb") as download_file:
    data = b"*" * 1000000
    with httpx.stream("POST", "https://httpbin.org/anything", data=data) as response:
        content_length = response.headers["Content-Length"]
        with tqdm(total=int(content_length)) as progress:
            for chunk in response.iter_bytes():
                download_file.write(chunk)
                progress.update(len(chunk))
```

Output:

```console
 32%|█████████████████████████                                                      | 317352/1000431 [00:00<00:00, 922202.25it/s]
```

**Comment by florimondmanca** [comment]
> you can't properly know how far you have progressed through the download because we'll often be performing gzip decompression

Ah, okay, I guess that's the reason why we wouldn't want users to use `tqdm` like I did above? Eg if the response is highly compressed, we'd call `.update()` with too many bytes compared to what was indicated by `Content-Length`, right?

**Comment by cdeler** [comment]
@tomchristie 
can we treat this issue as closed after #1268 ?

**Comment by tomchristie** [comment]
Yup! 👍👍👍
